### ISSUE TITLE
User Defined Networks

### ISSUE THREAD

**Issue by ktersius** [aws/amazon-ecs-agent#437]
I'm trying to see if [user defined networks ](https://docs.docker.com/v1.11/engine/userguide/networking/dockernetworks/#user-defined-networks) is supported? 

I've looked at the task definition options and could not find any place to set the network the container should connect to? 

Is it supported yet?

**Comment by samuelkarp** [comment]
User-defined networks are not yet supported on the task definition.  Can you help us understand what you'd intend to use them for?  Are you looking for something like service discovery, security isolation, or something else?

**Comment by ktersius** [comment]
Mostly in interested in the[ automated service discovery part](https://docs.docker.com/v1.11/engine/userguide/networking/dockernetworks/#docker-embedded-dns-server) where I can setup predefined domain names for me containers and connect my services via them. 

Currently I'm setting up a host DNS server which then scans the running containers and updates the DNS entries manually which is not ideal.

**Comment by stefansedich** [comment]
I am running into wanting this too for the service discovery aspect, I see it supports container links but I was of the understand that it was deprecated now in favor of using networks, is this something that will be implemented soon?

**Comment by mkleint** [comment]
my usecase is:
1. start selenium container
2. run an application build in another container and run the webdriver tests against the 'remote selenium' from the other container. --link doesn't seem to cut it as I not only need visibility from 2. to 1. but also from 1. to 2. (as the browser run in the selenium container needs to visit the app running in the other container.

**Comment by bheusinkveld** [comment]
my use case is I would like to be able to scale containers in services separately, but have communication to other containers in a different service/task definition.  If multiple different task definitions were able to connect to a user defined network, all containers across different task definitions would have network connectivity on that user defined network by hostnames.

**Comment by voltechs** [comment]
Would really love to see this. Currently Service Discovery is a huge pain requiring yet another service (which itself is usually cluster-based and self-discovers and then listens for other services). It's a messy solution, not to mention the Lambda "solutions" that are even more obnoxious to implement and maintain.

ECS needs native service discovery support out of the box. Specific example would be clustering services such as RabbitMQ or similar services.

**Comment by CpuID** [comment]
+1 to seeing this in place.

At a minimum passing through the equivalent of the `--network` docker run arg would be useful I think, defined in the container definition most likely.

**Comment by JBodkin** [comment]
I believe this needs to be looked into with a higher priority. The legacy links feature is currently deprecated and may be removed. This warning is in place on the documentation for the feature.

https://docs.docker.com/engine/userguide/networking/default_network/dockerlinks/

**Comment by techcto** [comment]
+1 - Really need this feature to create mysql replicas without putting on same host/task.

**Comment by eedwardsdisco** [comment]
+1 Linking is going away, and there are many services which require knowing their 'externally reachable host address (host ip + external port)' at runtime, which theoretically could be solved with user-defined networks

**Comment by lareeth** [comment]
I would very much like to be able to define my own network instead of being forced to use either 'Host', 'Bridge' or 'None'. The agent doesn't even need to create the network, just allow me to put in a network name that's custom and then at runtime see if it fails to start because the network doesn't exist.

**Comment by thovden** [comment]
I need to route traffic through a container that is running a VPN client. That way the actual containers can be used without modification when they need to use a VPN. Similar to the `--net=container:network` option that has removed from Docker.

**Comment by pseudobeer** [comment]
+1 Docker team seems to be planning links deprecation.

**Comment by alrooney** [comment]
👍  need this for elasticsearch nodes

**Comment by dzalbo** [comment]
👍 need this for hazelcast

**Comment by dbrainnet** [comment]
:+1: Would be useful for ZooKeeper.

**Comment by andreloc** [comment]
+ Consul ... service discovery

**Comment by bploetz** [comment]
👍 Use case for us is an nginx reverse proxy container which sits in front of an upstream API service running in another container on the same host. Currently our only option is using the deprecated link feature over the bridge network, or using something like DNS/ELB/Consul. But obviously we'd like to avoid making a network hop to call something that's running on the same host.

**Comment by adamrbennett** [comment]
A major disappointment I have with most (all?) orchestration tools is the assumption that all containers will be mapped to ports on the host. With overlay networks, this is not necessary. Containers can communicate within the network on ports that are not exposed or mapped to the host. This is clearly preferable as it almost completely eliminates any sort of port management and the possibility for port conflicts.

Start your containers in an overlay network, listen on standard ports (i.e.: 80/443) without worrying about conflicts, and setup a proxy to forward requests to your containers by name. Map your proxy to host port 80/443 and point your ELB at it. Manage it all using your service discovery DNS. This is the most elegant and maintainable solution, yet most orchestration tools will not support it. It's a crying shame. Literally, I am crying over it.

I shudder to think about managing 10,000 containers with port mapping. If each container exposes two ports, that's 20,000 ports I have to manage! Oh, I can make them map to random host ports, but now my proxy logic is so much more complicated, and someday I'll simply run out of ports. The bottom line is that a "scalable" solution that's built on port mapping is not scalable -- because mapping ports is not scalable.

I have modified the ECS agent to support this, and it works perfectly for my needs. However, it's less than ideal, because I lose the regular updates to the agent, unless I continually merge them in, and I have little to no visibility or control into the networks from the console or the CLI.

Guys, let's ditch the port mapping nonsense. It's not necessary with overlay networks.

**Comment by elasticsearcher** [comment]
@samuelkarp Is this currently in the works?

For anyone trying to do service discovery, take a look at the following article:
https://aws.amazon.com/blogs/compute/microservice-delivery-with-amazon-ecs-and-application-load-balancers/

From what I understand, you can use a single application load balancer to load balance up to 75 services by assigning a unique path prefix for each service, which you can then use to address your services. This doesn't cover all use cases, but should be enough for many applications.

**Comment by samuelkarp** [comment]
@elasticsearcher We're currently working on the ability to attach an ENI to a task and use native VPC networking.  We believe that this will address many of the use-cases described in this issue, as well as provide integration with existing VPC network topology and features that people are using today.

If you're interested in details, check out https://github.com/aws/amazon-ecs-agent/pull/701 (description of how we're planning to do this), [the `dev` branch of amazon-ecs-cni-pugins](https://github.com/aws/amazon-ecs-cni-plugins/tree/dev) (where we're working on some of the software behind this), as well as https://github.com/aws/amazon-ecs-agent/pull/777 and https://github.com/aws/amazon-ecs-agent/pull/779 (some of the changes necessary in the agent for this feature).

**Comment by TomaszSzawara** [comment]
👍   
simply exposing `--net=my-user-defined-network` in container definition, and adding user defined network in task definition is most appropriate. 

My use case assumes certain containers will join user defined networks and call each other by host. This setup is meant to run both outside of and inside AWS (through various development phases). No need to reinvent the wheel. Please support whats already there.

**Comment by mrt123** [comment]
👍
simply exposing --net=my-user-defined-network in container definition, and adding user defined network in task definition is most appropriate.

My use case assumes certain containers will join user defined networks and call each other by host. This setup is meant to run both outside of and inside AWS (through various development phases). No need to reinvent the wheel. Please support whats already there.

**Comment by johnbrandborg** [comment]
We require a number of containers to be bundled together with open communication, only exposing what needs to be consumed by the outside world.  Link is ugly and not scalable, and we need to be able to set the networks within out task definitions.  No need to over engineer whats already available.

**Comment by jamessewell** [comment]
Any updates here? - this is a really needed feature

**Comment by drzhouq** [comment]
This is much needed feature. I don't understand why AWS does not agree with the users. The use case is fairly common. Let's say you have a database container (serviceDB) that needs to be connected by multiple app containers (serviceApp). Put the database container and app container in one task definition and link them is not going to work.

**Comment by elasticsearcher** [comment]
Surprised no one's mentioned Weaveworks' integration with ECS, because it does pretty much what everyone here is asking for:
https://www.weave.works/docs/tutorials/old-guides/ecs/

Basically, Weave assigns an IP address to each container and runs an auto-managed DNS service, which lets any container in the same cluster address any other container by its name. The DNS service also automatically load-balances all containers.

I just tried it out and haven't encountered any issues so far. Just had to examine the ECS cluster setup script that they provide in the example to figure out the required SG and IAM configs.

Does anyone have experience with Weave and ECS? Any feedback would be super helpful.

**Comment by elasticsearcher** [comment]
@errordeveloper or @2opremio, would you mind chiming in please? I thought I'd loop you in since Weaveworks' solution seems to perfectly address this long-standing ECS feature request. Are there any limitations/concerns that we should be aware of or it's stable enough to use in production? :)

**Comment by 2opremio** [comment]
Yes, Weave Net should be able to solve most (if not all) the use cases presented above. It's production ready and we provide AMIs and Cloud Formation templates to run it.

See 

https://www.weave.works/docs/scope/latest/ami/
https://www.weave.works/docs/tutorials/old-guides/ecs/
https://www.slideshare.net/mobile/weaveworks/weaveworks-at-aws-reinvent-2016-operations-management-with-amazon-ecs

**Comment by elasticsearcher** [comment]
Thanks, @2opremio, that's great to hear! Weave Net makes connecting containerized apps so much easier.

**Comment by jamessewell** [comment]
That looks great in the interim - but it doesn't change the fact that ECS needs overlay networks if it wants to stay relevant.

**Comment by dario-simonetti** [comment]
I agree @jamessewell

**Comment by genexp** [comment]
We have a similar issue with eJabberd. We ended up deploying Kubernetes onto AWS using KOPS.  It makes all of this trivial.

**Comment by nick-wellinghoff-sp** [comment]
Another example use case. I am hosting a container which runs third party code where I want to restrict all outgoing routes. I lock down the container with iptables but lets say the third party code exploits a vuln and gains privileged escalation. Now they can override container iptables and get out to the net. If I had a user defined network that forced the container to use a restricted gateway I would not have to worry about the latest exploit taking down my whole stack. Put in the user defined networks por favor.

**Comment by dmulter** [comment]
+1 I just want to be sure that my ECS task links still work, or have an equivalent, when Docker eventually removes the deprecated legacy links feature.

**Comment by VimCommando** [comment]
+1 on this. I'm running a node.js application that uses a RabbitMQ service in a separate container. I can't figure out how to make the two containers talk to each other without knowing the IP addresses in advance; which makes no sense, because the IP addresses are assigned at container creation (both are running on the same host).

This was trivial using docker-compose inside an EC2 instance, but ECS doesn't seem to do it.

**Comment by alasdairrr** [comment]
+1 as well. Crazy this isn't supported :-O

**Comment by politician** [comment]
- Has anyone switched to something really similar to ECS like HashiCorp Nomad over this issue? How'd it go?

**Comment by boozedog** [comment]
+1

**Comment by nathanielks** [comment]
I'm interested in networks so that I can isolate tasks from each other. Per the docs:

> Containers that are collocated on a single container instance may be able to communicate with each other without requiring links or host port mappings. Network isolation is achieved on the container instance using security groups and VPC settings.

That's not ideal from a security perspective. By default, I don't want containers to be able to communicate with each other in production.

**Comment by juliaaano** [comment]
You might want to have a sidecar container running on the same network as the app container. The app should not be visible to anyone except the sidecar container. User defined networks is the right choice for that and docker compose / swarm already support it.

**Comment by nathanielks** [comment]
@juliaaano to clarify, we were planning on having the app code stored in a data-only container and then mount it as a volume to another container. Is this what you're referring to or something else?

**Comment by nathanielks** [comment]
(looking at these two articles: https://www.voxxed.com/2015/01/use-container-sidecar-microservices/ https://aws.amazon.com/blogs/compute/nginx-reverse-proxy-sidecar-container-on-amazon-ecs/)

**Comment by juliaaano** [comment]
@nathanielks that's one valid pattern. In my case I need a sidecar to do the SSL termination.

**Comment by elasticsearcher** [comment]
Task networking was finally released last week. The new awsvpc network mode lets you attach an ENI to each task, which makes them addressable by the DNS name of the ENI.

https://aws.amazon.com/about-aws/whats-new/2017/11/amazon-ecs-introduces-awsvpc-networking-mode-for-containers-to-support-full-networking-capabilities/

Thanks @samuelkarp for mentioning this being worked on a while back.

**Comment by nathanielks** [comment]
I saw that announcement and thought it was great! My main concern/question is surrounding the number of ENI's that can be attached to a single instance and how that relates to the number of containers you can launch on a single instance.

Let's take a `c5.xlarge` for example. It has 4 vCPU's and 4 attachable network interfaces. Each instance needs 1 interface to connect to the VPC, so that leaves us with 3 interfaces. Let's also say we're using an nginx container and set `worker_processes` to 1. In my mind, to take advantage of how nginx handles CPU processing, it would make sense to launch 3 nginx containers on the other 3 network interfaces with the thought that they would each get their own CPU core as well. This would also leave 1 core available for the system. Great! Everyone gets enough resources. This feels like I'm underutilizing a `c5.xlarge` though, to just run 3 nginx containers on a single instance.

Is my thinking flawed? Is there also enough room for other containers to be launched on the instance and not be bottlenecking the CPU because everyone is vieing for resources? I haven't found any good resources on how to size your containers so I'm searching in the dark.

**Comment by elasticsearcher** [comment]
@nathanielks Indeed, the limit on the number of ENIs you can attach to an instance is quite small, I didn't realize this was the case. As per the docs:

> Each task that uses the awsvpc network mode receives its own elastic network interface, which is attached to the container instance that hosts it. EC2 instances have a limit to the number of elastic network interfaces that can be attached to them, and the primary network interface counts as one. For example, a c4.large instance may have up to three elastic network interfaces attached to it. The primary network adapter for the instance counts as one, so you can attach two more elastic network interfaces to the instance. Because each awsvpc task requires an elastic network interface, you can only run two such tasks on this instance type. For more information about how many elastic network interfaces are supported per instance type, see IP Addresses Per Network Interface Per Instance Type in the Amazon EC2 User Guide for Linux Instances.

**Comment by adamrbennett** [comment]
Great point, @nathanielks -- unless there's more than meets the eye, the ENI limits are pretty much a non-starter for task networking.

My previous comment was a bit of a rant against port mapping, but my attitude has softened since then.

I've just submitted to the fact that I have to use host port mapping, and I'm fairly satisfied with the solution, given AWS will manage it for me with Target Groups and ELBv2. The issue of intra-host (and even inter-host) container communication is still there, but that can be addressed in a few different ways. This approach is not so bad if you are able to automate everything related to deployment, and the performance is better than you'd get with overlay networks.

My implementation is somewhat similar to this approach:
https://aws.amazon.com/blogs/compute/service-discovery-an-amazon-ecs-reference-architecture/

I'm using ELBv2 to avoid having so many ELBs, and if you use an automation tool (e.g.: Terraform) then the Lambda for service registration is also not necessary. Without these adjustments I wouldn't be fully satisfied, but with them (and automation) it is a totally manageable solution. Complete automation would be more difficult (or even impossible) if you're not using Route53.

In any case, support for overlay networks is still necessary as container links are deprecated. Given the Docker API and the Go docker client already support it, there really shouldn't be much trouble to implement it (barring edge cases, potential security concerns, etc).

**Comment by elasticsearcher** [comment]
FYI, AWS Route 53 just released the Auto Naming API for Service Name Management and Discovery:
https://aws.amazon.com/about-aws/whats-new/2017/12/amazon-route-53-releases-auto-naming-api-name-service-management/

**Comment by softwareplumber** [comment]
Just been experimenting with ENI networking for containers. Not wildly impressed; overly complex for my simple use cases. Add me to the list of people who would prefer to use user defined networks.

**Comment by vbarba** [comment]
Yes please. Add me to the list. It is not possible to link/volumes_from in awsvpc and this is a problem for us.

**Comment by triesz** [comment]
+1 use case: I am trying to start [reportportal](http://reportportal.io/docs/Deploy-ReportPortal) on AWS (which is using consul) and without this feature a simple 30 minutes task is exploding badly

**Comment by samitpal** [comment]
+1 for this exposing --net=my-user-defined-network in container definition

**Comment by antoiner77** [comment]
+ 1 this can really be useful

**Comment by lnehrin** [comment]
Weave panics the 4.14.33-51.34.amzn1.x86_64 kernel (Amazon Linux AMI 2018.03) forcing us to roll back (painfully) to the previous kernel.     That prompted me to look at docker swarm, which works fine on that kernel, but the lack of a "--network" flag support by ECS brings us to a decision of whether to dump ECS entirely.   The VPC network solution is interesting from a security perspective for externally exposed ports, but doesn't adequately provide what most of us are asking for:  an overlay network for containers on different ECS instances to be able to talk to each other.   Weave can do that, but a bit of a hassle to configure compared to docker swarm.   I wanted a light weight solution instead of being forced to look at DCOS or Kubernetes or even PCF - which all look heavy when all I want is an overlay network.

**Comment by tdb-alcorn** [comment]
+1

**Comment by arlobryer** [comment]
Hi there, is there any update on being able to use user defined networks on ECS?

It seems like this is pretty essential to the service being friendly to use/in line with Docker recommendations for communication across containers.

**Comment by Cephei** [comment]
Please add this functionality. Spent two days trying to get `awsvpc` to work. Can't believe this is so difficult getting two containers to talk to each other in ECS. As a non-expert in IT/networking, ECS documentation is severely lacking in its ease of adoption. Additionally, I would rather not be forced to use & pay for VPC/NAT Gateways just to get containers to talk. Is there any way around this?!?

**Comment by dschaaff** [comment]
Service discovery works for bridged network tasks now I believe. But using load balancers is the best way to handle inter service communication in ECS right now.

**Comment by Cephei** [comment]
Three days spent trying to get `awsvpc` to work and nothing is any more clear than when I started. It took me all of 10 minutes to run a private user-defined-network on my local machine with Akka remoting, only exposing ssl/websockets to the host, using docker-compose.

As far as I can tell, the AWS docker rabbit-hole: an ECS cluster with `awsvpc` & service discovery requires...
1. Route 53 private hostname (configured separately and that you can only delete via CLI for some reason)
2. Load balancer / Target group for networking into the container? (even though I don't want this yet)
3. Two subnets; one public / one private?
4. NAT Gateway to reach the internet
5. Security group for each ECS service (this was a nice feature)

I don't even know if this really outlines everything required. I can't tell if it's correct as I wasn't able to get it working. All this seems to offer is force me to use more AWS services in a design that's un-necessarily complex. I think I should just manage docker myself. ECS needs some serious UX love to make things easier.

**Comment by 2opremio** [comment]
Just use Weave Net

**Comment by rtasson** [comment]
Is there any update on user defined networking with ECS? With the advent of daemon-style tasks (https://aws.amazon.com/about-aws/whats-new/2018/06/amazon-ecs-adds-daemon-scheduling/) it would be nice to use user-defined networking to allow simple service discovery; you can count on containers being deployed on each instance, and simply communicate with their hostname. This solution is elegant and simple; every other solution is more engineering, more services to monitor and deploy. The next best solution is container linking, a deprecated feature. We have already run into ENI limits on instances. I'm astounded this issue is nearly 2 years old with no update besides "use a third party solution."

**Comment by Cephei** [comment]
I honestly moved to Azure AKS because of this issue. Got it working within 30min - although I had to learn a bit of Kubernetes. 

The ENI constraints are absurd. Perhaps try looking into Amazon EKS. Although they charge per-hour for each cluster.

**Comment by bploetz** [comment]
This issue has been open for two years, and is the second most commented open issue. The silence from the AWS is deafening. Can we please get some kind of an update?

**Comment by lnehrin** [comment]
We use Weave with ECS..... *but* there was recently an issue with a kernel
patch on Amazon Linux (right about the time 2018-03 was released) where the
kernel patch was broken such that weave would panic the kernel - resulting
in ground-hog day style continuous rebooting.

After reverting my instances using snapshots with the older kernel, I had
to wait a couple weeks for Amazon to provide a kernel patch that would work
with Weave.  In the mean time, I experimented with docker swarm and was
impressed with how well the overlay networks simply work.....it just didn't
work with ECS.  I almost got to the point of experimenting with ECS
instances custom built on Ubuntu linux.

It sure seams that Amazon is purposely ignoring supporting overlay networks
in favor of their solution using additional network interfaces on instances
and the stuff associated with that  (and the unfortunate issue that
instances are limited to a rather small number of ENIs).     It's really
making me want to take a serious look at Kubernetes and see whether EKS is
worth learning or not and move away from ECS issues.

**Comment by softwareplumber** [comment]
Obviously Amazon have  a right to pursue whatever commercial strategy they like, but I think you are right about their reasons for ignoring this.

It's a shame. I agree that Weave, EKS, or even the (needlessly complex IMO) Amazon solution might be the right way to deploy at scale. But for those of us who target smaller scale systems or just want to migrate dev or POC system to AWS, there's no question that overlay networks would make it much simpler to port a small scale solution (...that might already be running on some developer's laptop) to AWS.

I too am on the point of abandoning ECS in favor of Kubernetes... when I have the time [yeah, right]

**Comment by Cephei** [comment]
I've spent a couple weeks now working with Kubernetes. I'm very happy with Azure's ecosystem: using VS Code for all development, ACR (container rep), AKS (k8 service). I use docker-compose for local container testing and kubectl to deploy my kubernetes yaml configuration to Az. It's extremely easy with AKS although documentation could always be better.

I wish I had looked at Amazon's Kubernetes services earlier as I might not have switched providers; thus I can't give you a well-rounded comparison. It looks like Kubernetes is the future though. Azure also had something called `Container Services` - similar to ECS, but they've rebranded as `AKS` and advocate strongly for K8 over competing orchestration libraries. I believe AWS is doing the same, but have nothing to back this up officially. (would be nice to see an official roadmap for ECS vs EKS)

Regardless of cloud provider you use, I really recommend spending some time learning K8. There's a slight learning curve but 1-2 weeks can get you running. I have an Akka cluster of containers deployed and conceptually, I could take the same K8 configuration file and apply it to AWS with minor changes of switching container registries. Although I do know AWS has some K8-specific annotations you need to be aware of.

**Comment by wyqydsyq** [comment]
I've spent the last week banging my head against the wall with this and have yet to find a good solution.

1) As previously mentioned, Docker's `--link` (and by extension, the task definition "`link`" property) are being deprecated, additionally for it to work this requires two containers to be in the same task definition (so they are placed on the same host and share the same bridge network) so it isn't scalable nor future-proof

2) I've tried using the R53/ECS Service Discovery / Auto naming with no luck - I get the appropriate R53 records created, pointing to the correct container and port, but calling the generated endpoint URL from a container in the same VPC resulted in it not being able to resolve the hostname, as if the container is not aware of the private R53 zone

3) All other solutions I can find involve feel hacky and non-intuitive, including adding third-party discovery services like HashiCorp Consul

With Amazon's apparent decision to not support overlay networks, lack of adequate documentation and resources on using their Service Discovery Service outside setting it up via the console's GUI, and how comparatively simple networking appears to be via Docker Swarm or Kubernetes, ECS is becoming an increasingly unappealing solution.

**Comment by dzalbo** [comment]
@wyqydsyq the ECS Service Discovery works pretty well for me (apart from the fact the CloudFormation side of it is still somewhat shaky)... are you sure you do not have some custom DNS settings in the host ec2 instance or the container itself?

**Comment by pecigonzalo** [comment]
The problem with ECS SD, as with EKS is that they are not released to all regions.

**Comment by psyhomb** [comment]
Anything new on this?

Here's the doc just for reference:
https://docs.docker.com/network/bridge

**Comment by galindro** [comment]
I need this feature too... awsvpc mode, IMO, isn't good enough because the instances ENI attachment limitations.

**Comment by sGeeK44** [comment]
If someone is interrested, i have found a workaround....

Create a  script setup-server.py:
```
import os
import docker

docker_client = docker.from_env()
networks_client = docker_client.networks
network_name = os.environ['CONTAINER_NETWORK']
network_list = networks_client.list(network_name)
if network_list:
    network = network_list[0]
else:
    network = networks_client.create(network_name, driver="bridge")

container_id = os.environ['HOSTNAME']
network.connect(container_id)
```

Create a  script run-server.sh:
```
#!/bin/bash

python3 ./setup-server.py

# Add your usaul entry point there
.....
```

In dockerfile add scripts and python tools
```
RUN apt-get update \
    && apt-get --yes install \
        python3-pip

RUN pip3 install docker

COPY setup-server.py setup-server.py
RUN chmod +x setup-server.py

COPY runserver.sh runserver.sh
RUN chmod +x runserver.sh
```

In ECS task:
* add env var for user defiine network name
* add shell script in entry point  ["/bin/bash","/your/path/runserver.sh"]
* Shared a volume to access at host docker deamon. /var/run/docker.sock:/var/run/docker.sock

It works nicely for me. But I hope ECS Team will make something to get user-defined network functionnality.

**Comment by dzalbo** [comment]
@psyhomb https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-discovery.html

**Comment by psyhomb** [comment]
I tought @yunhee-l was talking about some new features regarding service discovery, but obviously I was wrong, I'm very familiar with service discovery (and doc) they have launched some time ago but they're forcing us to use Route53 so that's why I don't what to use it.
This is where the answer to the question, regarding user-defined networks and why they're still "working" on the implementation, is hiding.

**Comment by key-er** [comment]
Use case is nginx for proxying request to different routes that are different containers in a custom network. 

For fargate it says:
> host port mappings are not valid when the network mode for a task definition is host or awsvpc 
>> that is the only available option I saw in drop down).

> To specify different host and container port mappings, choose the Bridge network mode 
>> bridge mode is only for Windows host
